.github/workflows/release.yml

```diff
@@ -1,63 +1,112 @@
+          CIBW_SKIP: "*-win32 *-musllinux*"
+          CIBW_ARCHS_MACOS: x86_64 arm64
+  build_wheels:
+    name: Build ${{ matrix.os }} ${{ matrix.only }}
+    needs: generate-wheels-matrix
+    strategy:
+      matrix:
+        include: ${{ fromJson(needs.generate-wheels-matrix.outputs.include) }}
+    runs-on: ${{ matrix.os }}
     defaults:
       run:
-        shell: bash -l {0}
-
+        shell: bash
     steps:
       - uses: actions/checkout@v3
-      - uses: conda-incubator/setup-miniconda@v2
         with:
-          activate-environment: test-environment
-          python-version: ${{ matrix.python-version }}
-          channels: conda-forge/label/testing,conda-forge
+          # We need the full history to generate the proper version number
+          fetch-depth: 0
 
-      - name: Install dependencies
-        run: |
-          PACKAGES="cython fiona matplotlib-base numpy pyproj pykdtree scipy"
-          PACKAGES="$PACKAGES owslib pep8 pillow pyshp pytest"
-          PACKAGES="$PACKAGES pytest-xdist setuptools_scm shapely"
-          conda install $PACKAGES
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.10'
+
+      - uses: pypa/cibuildwheel@v2.13.1
+        with:
+          only: ${{ matrix.only }}
+
+      - uses: actions/upload-artifact@v3
+        with:
+          path: ./wheelhouse/*.whl
+
+  build_sdist:
+    name: Build source distribution
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          # We need the full history to generate the proper version number
+          fetch-depth: 0
 
-      - name: Create sdist
-        run: python setup.py build_ext sdist
+      - uses: actions/setup-python@v4
+        name: Install Python
+        with:
+          python-version: '3.11'
 
-      - name: Save built packages as artifact
-        uses: actions/upload-artifact@v3
+      - name: Build sdist
+        run: |
+          python -m pip install build
+          # install compiles the cython *.pyx to *.cpp for us
+          python -m pip install .
+          python -m build --sdist
+      - uses: actions/upload-artifact@v3
         with:
-          name: packages-${{ runner.os }}-${{ matrix.python-version }}
-          path: dist/
-          if-no-files-found: error
-          retention-days: 5
+          path: dist/*.tar.gz
 
   publish:
     name: Publish to PyPI
-    needs: build
+    needs: [build_wheels, build_sdist]
     environment:
       name: PyPI
       url: https://pypi.org/project/cartopy
     permissions:
       id-token: write  # IMPORTANT: this permission is mandatory for trusted publishing
-    runs-on: ubuntu-18.04
+    runs-on: ubuntu-latest
 
     steps:
       - name: Download packages
         uses: actions/download-artifact@v3
-
-      - name: Consolidate packages for upload
-        run: |
-          mkdir dist
-          cp packages-*/* dist/
```

Comment:
Are the sdist and all the wheels getting put in a single directory (or otherwise where they can be found for upload) without this?

Reply:
Yes, they are getting put into one large artifact that you'd have to download. See here for a CI run testing on my fork to see what that looks like: https://github.com/greglucas/cartopy/actions/runs/5371702871

For the black magic, I was relying on these links: https://github.com/pypa/cibuildwheel/discussions/1261
https://iscinumpy.dev/post/cibuildwheel-2-10-0/#only-210

Reply:
Ah, I see what's going on now with the downloading of artifacts. Nice and clean.